core/src/main/java/com/linecorp/armeria/client/circuitbreaker/CircuitBreakerClient.java

```diff
@@ -118,27 +71,33 @@ public O execute(ClientRequestContext ctx, I req) throws Exception {
         }
 
         if (circuitBreaker.canRequest()) {
-            final O response;
-            try {
-                response = delegate().execute(ctx, req);
-            } catch (Throwable cause) {
-                circuitBreaker.onFailure(cause);
-                throw cause;
-            }
-
-            response.completionFuture().handle(voidFunction((res, cause) -> {
-                // Report whether the invocation has succeeded or failed.
-                if (cause == null) {
-                    circuitBreaker.onSuccess();
-                } else {
-                    circuitBreaker.onFailure(cause);
-                }
-            })).exceptionally(CompletionActions::log);
-
-            return response;
+            return doExecute(ctx, req, circuitBreaker);
         } else {
             // the circuit is tripped; raise an exception without delegating.
             throw new FailFastException(circuitBreaker);
         }
     }
+
+    /**
+     * Invoked when the {@link CircuitBreaker} is in closed state.
+     */
+    protected abstract O doExecute(ClientRequestContext ctx, I req, CircuitBreaker circuitBreaker)
+            throws Exception;
+
+    /**
+     * Reports a success or a failure to the specified {@link CircuitBreaker} according to the completed value
+     * of the specified {@code future}. If the completed value is {@code null}, this doesn't do anything.
```

Comment:
Just curious. Is it any special reason of adding `null` as a return value? Because it seems that there were only success and failure before.